### PR TITLE
replication: reduce RPC calls when VR state is primary

### PR DESCRIPTION
**Pull request by yati1998**
Avoid extra RPC calls as request will be requested again for updating the LastSyncTime in the status. The image need to be promoted only one time not always during the reconcile.

fixes: #250

Co-authored-by: Madhu Rajanna <madhupr007@gmail.com>
Signed-off-by: yati1998 <ypadia@redhat.com>

### CODE REVIEW

**Comment by yati1998** [comment]
@Madhu-1 @ShyamsundarR

File: controllers/replication.storage/volumereplication_controller.go [hunk 1]
```diff
@@ -273,7 +273,24 @@ func (r *VolumeReplicationReconciler) Reconcile(ctx context.Context, req ctrl.Re
 
 	switch instance.Spec.ReplicationState {
 	case replicationv1alpha1.Primary:
-		replicationErr = r.markVolumeAsPrimary(vr)
+		// Avoid extra RPC calls as request will be requested again for
+		// updating the LastSyncTime in the status. The image need to be
+		// promoted only one time not always during the reconcile.
```
Comment (Madhu-1) [inline controllers/replication.storage/volumereplication_controller.go:278]:
```suggestion
// updating the LastSyncTime in the status. The volume needs to be
// Replication enabled and promoted only once, not always during each reconcilation.
```

**Review by Madhu-1 — COMMENTED** [review]
LGTM, let's wait for the test results. If everything is good, we can merge this one.

File: controllers/replication.storage/volumereplication_controller.go [hunk 2]
```diff
@@ -255,25 +255,30 @@ func (r *VolumeReplicationReconciler) Reconcile(ctx context.Context, req ctrl.Re
...
 	var requeueForResync bool
 
 	switch instance.Spec.ReplicationState {
 	case replicationv1alpha1.Primary:
-		replicationErr = r.markVolumeAsPrimary(vr)
+		// Avoid extra RPC calls as request will be requested again for
+		// updating the LastSyncTime in the status. The volume needs to be
+		// replication enabled and promoted only once, not always during
+		// each reconciliation.
+		if instance.Status.State != replicationv1alpha1.PrimaryState {
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply controllers/replication.storage/volumereplication_controller.go:267]: Yes, we can do that if it avoids corner cases, but here is the `instance.Status.ObservedGeneration` wont be equal to the `instance.Generation` as the `status.ObservedGeneration` is not updated yet right? let's take it in follow-up PR as it also requires extensive testing and also this PR is already getting tested for some time. @ShyamsundarR WDYT?

[the remark this answers is not in the document]
Comment (ShyamsundarR) [reply controllers/replication.storage/volumereplication_controller.go:267]: In this case the instance.Generation will equal the instance.Status.ObservedGeneration as there is no change to spec. The reconcile only updates the status for the sync time, which will change the revision.

I am fine taking it into another PR as the generation match is not tested in other conditions as well.

[the remark this answers is not in the document]
Comment (Madhu-1) [reply controllers/replication.storage/volumereplication_controller.go:267]: Yes, correct. Let's take it as a follow-up PR, as we need to test it verywell.

**Review by Madhu-1 — APPROVED** [review]

**Pull request merged**